google/cloud/spanner/internal/clock.h

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_INTERNAL_CLOCK_H
+#define GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_INTERNAL_CLOCK_H
+
+#include "google/cloud/spanner/version.h"
+#include <chrono>
+#include <mutex>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+/**
+ * A simple `Clock` class that can be overridden for testing.
+ *
+ * All implementations of this class are required to be thread-safe.
+ */
+class Clock {
+ public:
+  using time_point = std::chrono::steady_clock::time_point;
+  using duration = std::chrono::steady_clock::duration;
+
+  virtual ~Clock() = default;
+  virtual time_point Now() = 0;
```

Comment:
As discussed offline, I think `Now()` is logically a const operation as it's not modifying any logical state of the clock instance. It's simply reading the clocks state. If the clocks state is changed between successive calls of `Now()`, that's simply because "something else" mutated the clock's state. That has nothing to do with what `Now()` does.

That said, I know this may be a controversial opinion. I'm certainly not going to block this PR on my opinion. I'm happy to defer to @mr-salty or @devbww on this point.